### ISSUE TITLE
Moffat1D and Moffat2D derivatives

### ISSUE THREAD

**Issue by ycopin** [astropy/astropy#8094]
The derivatives in Moffat1D are wrong:

https://github.com/astropy/astropy/blob/2b8fe1305b2689d8b2715f98985e36f2c7d5c0d9/astropy/modeling/functional_models.py#L2223

They should be replaced by:
```
    @staticmethod
    def fit_deriv(x, amplitude, x_0, gamma, alpha):
        """One dimensional Moffat model derivative with respect to parameters"""

        fac = (1 + (x - x_0) ** 2 / gamma ** 2)
        d_A = fac ** (-alpha)
        d_x_0 = (2 * amplitude * alpha * (x - x_0) * d_A / (fac * gamma ** 2))
        d_gamma = (2 * amplitude * alpha * (x - x_0) ** 2 * d_A /
                   (fac * gamma ** 3))
        d_alpha = -amplitude * d_A * np.log(fac)
        return [d_A, d_x_0, d_gamma, d_alpha]
```

In Moffat2D:

https://github.com/astropy/astropy/blob/2b8fe1305b2689d8b2715f98985e36f2c7d5c0d9/astropy/modeling/functional_models.py#L2301

only derivative wrt gamma is wrong. It should be
```
        d_gamma = 2 * amplitude * alpha * d_A * rr_gg / ((1 + rr_gg) * gamma ** 3)
```

**Comment by pllim** [comment]
These were originally added by @adonath as part of #1251 and has since been renamed a few times.

**Comment by adonath** [comment]
I remember that for the `Beta1D` and `Beta2D` models I computed the derivatives and reference values in a notebook using using Sympy (http://nbviewer.jupyter.org/gist/adonath/6088712). So I'm pretty sure it used to be correct at that time. Maybe when the rename to `Moffat1D` and `Moffat2D` happened, the parametrization changed as well and the derivatives where not adapted? From a quick look  in the code it seems there are no unit tests for the model derivatives.

@ycopin I didn't check whether your proposed formulae are correct, please check against Sympy or some other algebra library and post the result as a reference. In case you still find the current formulae are incorrect, would you mind preparing a minimal pull request with the correct ones? Ideally one should add unit tests for the model derivatives, but this is a somewhat larger project...

**Comment by ycopin** [comment]
I used Wolfram Alpha for the derivatives, but I trust the derivatives in the aforementioned notebook are correct. The issue is that they were not properly implemented in the code (in particular the `d_A ** alpha` terms, where it should just be `d_A`). 

For sure, the derivatives are currently incorrect, and the fit does not converge if not using ad-hoc numerical derivative estimates. New derivatives solve the issue (at least in my 1D case). 

I'll try to prepare a PR, but that might take more time...

**Comment by ycopin** [comment]
BTW, I wonder why this was not detected before, as it seems there are some derivative tests in modeling.

**Comment by chris-simpson** [comment]
You got this one wrong. The derivative in Moffat2D was correct and the PR made it wrong. There's only a single power of gamma in the denominator, not 3, because 2 of them have been subsumed into `rr_gg`. I'll use Wolfram alpha for the calculus so you don't have to check mine...
![Screen Shot 2020-09-29 at 10 59 20 AM](https://user-images.githubusercontent.com/22304216/94620031-a3dd7580-0249-11eb-8d54-95da5c21e3b4.png)
<img width="396" alt="Screen Shot 2020-09-29 at 11 45 51 AM" src="https://user-images.githubusercontent.com/22304216/94620059-ae980a80-0249-11eb-982f-f14973fe1a55.png">

**Comment by saimn** [comment]
Finding the same issue with `d_gamma` with Sympy:
![image](https://user-images.githubusercontent.com/311639/94623734-4702af00-028b-11eb-8c18-db81e6ac9cf8.png)

Full notebook here: https://gist.github.com/saimn/f79197fe45bb96d90f6ca3fa54910ac2
The other fixes (implemented in #8108) for Moffat1D are correct.

**Comment by ycopin** [comment]
I'm sorry I introduced a bug in Moffat2D derivatives... :-/ As mentioned earlier, verifying the formal derivatives using e.g. Wolfram is not necessarily sufficient, but one should check the actual implementation, thanks @saimn for doing so (even if the tested implementation might not be the most efficient). I still wonder why this implementation error was not caught by a test comparing formal and numerical derivatives...
